aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -0,0 +1,1211 @@
+
+                        // The current logic is to just ask caller to speculatively abort.
+                        // TODO: complex handling (e.g. scheduler maintenance), also revisit Wait.
+                        return Ok(false);
+                    }
+
+                    ret
+                },
+                (Low, Some(false)) | (Medium, Some(false)) => {
+                    // There is value in finishing the speculative execution, as the transaction
+                    // has not been previously aborted due to this dependency: proceeding allows
+                    // an optimistic approach, and recording the aborted dependency upon failure.
+                    DependencyResolution::None
+                },
+                (High | Medium | Low, None) | (Highest, Some(_)) => {
+                    unreachable!("abort_likely checked for high / medium / low priority")
+                },
+            };
+
+            match resolution {
+                DependencyResolution::SafeToProceed | DependencyResolution::None => {
+                    return Ok(true);
+                },
+                DependencyResolution::Wait(dep_condition) => {
+                    Self::wait(dep_condition);
+                    // Next iteration of the loop will try to resolve again.
+                },
+                DependencyResolution::Halted => {
+                    return Ok(false);
+                },
+            }
+        }
+    }
+
+    pub(crate) fn finish_execution(
+        &self,
+        txn_idx: TxnIndex,
+        incarnation: Incarnation,
+        invalidated_versions: BTreeSet<(TxnIndex, Incarnation)>,
+    ) -> Result<(), PanicError> {
+        if incarnation > 0 {
+            // TODO: make sure we don't kill switch 0-th incarnations.
+            // Record aborted dependencies. Only recording for incarnations > 0 is in line with the
+            // optimistic value validation principle of Block-STMv2. 0-th incarnation might invalidate
+            // due to the first write, but later incarnations could make the same writes - in which case
+            // there is no need to record (and stall, etc) the corresponding dependency.
+            self.aborted_dependencies[txn_idx as usize]
+                .lock()
+                .record_dependencies(invalidated_versions.iter().map(|(idx, _)| *idx));
+        }
+
+        let mut propagation_queue = BTreeSet::from([txn_idx]);
+        for (invalidated_idx, invalidated_incarnation) in invalidated_versions {
+            let mut to_propagate = self.try_abort(invalidated_idx, invalidated_incarnation)?;
+            propagation_queue.append(&mut to_propagate);
+        }
+        if self.txn_status[txn_idx as usize].finish_execution(incarnation)? {
+            // After updating the status, check if more transactions can be committed. Important to
+            // be called after finish execution.
+            self.try_commit(txn_idx)?;
```

Comment:
there seems some hidden invariant that ensures the invalidation covers everything, probably need to document somewhere i.e. if a txn B reads a wrong value from incarnation X, it's guarantees that X+1 will invalidate it. it's less intuitive than the pull based validation

Reply:
Would the top-level comment in SchedulerV2 be a good place for explaining these?